[BUG] When CoroutineMock object raised an Exception...

[ python 3.6 + asynctest 0.12.0 ]

```python
import asynctest

class MyException(BaseException):
     pass

class MyTestCase(asynctest.TestCase):
    @staticmethod
    async def test_await():
        mock = asynctest.CoroutineMock()
        mock.side_effect = MyException

        mock.assert_not_called()
        mock.assert_not_awaited()

        try:
            await mock()
        except MyException:
            pass

        mock.assert_called_once()
        mock.assert_awaited_once()  # FAILED,  AssertionError: Expected 'mock' to have been awaited once. Awaited 0 times.
```

as you've been able to notice the "FAILED".

--- a/asynctest/mock.py
+++ b/asynctest/mock.py
@@ -540,30 +540,31 @@ class CoroutineMock(Mock):
     def _mock_call(_mock_self, *args, **kwargs):
         try:
             result = super()._mock_call(*args, **kwargs)
-            _call = _mock_self.call_args
-
-            @asyncio.coroutine
-            def proxy():
-                try:
-                    if _isawaitable(result):
-                        return (yield from result)
-                    else:
-                        return result
-                finally:
-                    _mock_self.await_count += 1
-                    _mock_self.await_args = _call
-                    _mock_self.await_args_list.append(_call)
-                    yield from _mock_self.awaited._notify()
-
-            return proxy()
         except StopIteration as e:
             side_effect = _mock_self.side_effect
             if side_effect is not None and not callable(side_effect):
                 raise
 
-            return asyncio.coroutine(_raise)(e)
+            result = asyncio.coroutine(_raise)(e)
         except BaseException as e:
-            return asyncio.coroutine(_raise)(e)
+            result = asyncio.coroutine(_raise)(e)
+
+        _call = _mock_self.call_args
+
+        @asyncio.coroutine
+        def proxy():
+            try:
+                if _isawaitable(result):
+                    return (yield from result)
+                else:
+                    return result
+            finally:
+                _mock_self.await_count += 1
+                _mock_self.await_args = _call
+                _mock_self.await_args_list.append(_call)
+                yield from _mock_self.awaited._notify()
+
+        return proxy()
 
     def assert_awaited(_mock_self):
         """

--- a/test/test_mock.py
+++ b/test/test_mock.py
@@ -367,6 +367,14 @@ class Test_CoroutineMock_awaited(asynctest.TestCase):
 
         self.assertEqual(mock.await_count, 1)
 
+        mock.reset_mock()
+        mock.side_effect = RuntimeError
+
+        with self.assertRaises(RuntimeError):
+            yield from mock()
+
+        self.assertEqual(mock.await_count, 1)
+
     @asyncio.coroutine
     def test_awaited_from_autospec_mock(self):
         mock = asynctest.mock.create_autospec(Test)

